### PR TITLE
Malte lig 4894 fix gatherlayer

### LINKED ISSUE #1528: Bug in `GatherLayer.backward`
Hi,

We've been implementing a model at [cellarium-ml](https://github.com/cellarium-ai/cellarium-ml) using your `NTXentLoss`. Comparing the model training with a single GPU and two GPUs we noticed that they do not match. By investigating it we found an apparent bug in the `GatherLayer.backward` where gradients are not sum-reduced over GPUs. Here is our fixed version (https://github.com/cellarium-ai/cellarium-ml/blob/main/cellarium/ml/distributed/gather.py#L17-L21):

```py
    @staticmethod
    def backward(ctx, *grads) -> torch.Tensor:
        grad_out = grads[dist.get_rank()].contiguous()
        dist.all_reduce(grad_out, op=dist.ReduceOp.SUM)
        return grad_out
```

and the [test](https://github.com/cellarium-ai/cellarium-ml/blob/main/tests/distributed/test_gather.py) we wrote. Would you agree that this is indeed a bug? I would be happy to contribute a PR with the fix.

### PR DESCRIPTION
**Pull request by MalteEbner**
closes #1528

## Description

Fixes the `GatherLayer` by using the implementation from solo learn.

## Tests
Adds a test for the `GatherLayer` by testing a model using the `NTXentLoss` criterion. It compares the training behaviour for these two cases and ensures that it is exactly the same:

1. n_devices=1, batch_size=8
2. n_devices=2, batch_size=4

The test is in the new file `test_dist__gather.py`. This is needed, because using a DDPStrategy causes the file to be executed once per device.
Before the fix, the test failed.

## Next tests

This test only tests the `NTXentLoss` criterion, the other models need to be tested as well.

## Testing the full SimCLR model

I also tried to have similar test when using a SimCLR model. However, it is extremely hard to get exactly the same training when using it.

### Randomness causes different behaviour between n_devices=2 and n_devices=1

Results:

Using the SimCLR transform leads to different behaviour when using n_devices=2 compared to only 1 device. Even seeding does not help. This is caused by the different number of samples and thus the different random seeds. E.g.
- n_devices = 1, batch_size=8 processes the samples in order 0, 1, ... 7 and thus uses seed_0, seed_1, .... seed_7. Thus sample_1 uses seed_1
- n_devices = 2, batch_size=4 processes the samples in order 0, 2, 4, 6 (device 0) and 1, 3, 5, 7 (device 1). Each of them uses seed_0, seed_1, ... in parallel, as these are two different process with their own seeding each. Thus sample_1 gets seed_0, which makes it differ.

Thus only removing randomness makes the output of the dataloader the same for the n_devices=2 and n_devices=1 cases.

The same problem also applies to any randomness in the model itself, e.g. in dropout layers.

### Batch normalization causes different behaviour between n_devices=2 and n_devices=1

Batch normalization or any other operation using information from other samples in the same batch behaves differently when using `n_devices=2 & batch_size=4` compared to `n_devices=1 & batch_size=8`. The batch normalisation would need to be synchronised as well for this to work. 
As pointed out by Guarin, we could use SyncBatchNorm to avoid this: https://lightning.ai/docs/pytorch/stable/common/trainer.html#sync-batchnorm

### CODE REVIEW

**Review by guarin — APPROVED** [review]
LGTM!

Let some comments regarding typos but otherwise everything looks good :)

File: lightly/utils/dist.py [hunk 1]
```diff
@@ -8,22 +8,21 @@ class GatherLayer(torch.autograd.Function):
...
         dist.all_gather(output, input)
         return tuple(output)
 
     @staticmethod
-    def backward(ctx, *grads: torch.Tensor) -> torch.Tensor:
-        (input,) = ctx.saved_tensors
-        grad_out = torch.empty_like(input)
-        grad_out[:] = grads[dist.get_rank()]
+    def backward(ctx, *grads) -> torch.Tensor:
+        all_gradients = torch.stack(grads)
```
Comment (ordabayevy) [inline lightly/utils/dist.py:23]: The code here can be optimized by avoiding stacking (see https://github.com/cellarium-ai/cellarium-ml/blob/220ba90b47378c99d4c08b9d91c5c31b796cb3ca/cellarium/ml/distributed/gather.py#L17-L21):

```py
    @staticmethod
    def backward(ctx, *grads) -> torch.Tensor:
        grad_out = grads[dist.get_rank()].contiguous()
        dist.all_reduce(grad_out, op=dist.ReduceOp.SUM)
        return grad_out
```

Comment (MalteEbner) [reply lightly/utils/dist.py:23]: The code you proposed causes the test to fail. I have tested it using two devices and by calling `TestGatherLayer().test(None)`. Thus I would rather keep the existing code, as it is the only one causing the training to come to exactly the same values.

### the code you proposed:
```
grads: (tensor([[-0.0576, -0.0557, -0.0831, -0.0729],
        [-0.0867, -0.0694, -0.0778, -0.0508],
        [-0.0852, -0.0687, -0.0781, -0.0520],
        [ 0.0289, -0.0130, -0.0879, -0.1265]]), tensor([[0.0354, 0.0653, 0.0552, 0.0957],
        [0.0354, 0.0651, 0.0548, 0.0950],
        [0.0353, 0.0649, 0.0546, 0.0946],
        [0.0352, 0.0647, 0.0542, 0.0939]]))
grads: (tensor([[0.0454, 0.0753, 0.0521, 0.0859],
        [0.0457, 0.0759, 0.0524, 0.0865],
        [0.0457, 0.0759, 0.0524, 0.0865],
        [0.0393, 0.0649, 0.0451, 0.0740]]), tensor([[-0.0629, -0.0525, -0.0833, -0.0704],
        [-0.0692, -0.0552, -0.0817, -0.0648],
        [-0.0726, -0.0567, -0.0807, -0.0617],
        [-0.0773, -0.0587, -0.0794, -0.0574]]))
grad_out: tensor([[-0.1205, -0.1082, -0.1664, -0.1434],
        [-0.1559, -0.1246, -0.1595, -0.1156],
        [-0.1579, -0.1254, -0.1588, -0.1136],
        [-0.0484, -0.0718, -0.1674, -0.1839]])
grad_out: tensor([[-0.1205, -0.1082, -0.1664, -0.1434],
        [-0.1559, -0.1246, -0.1595, -0.1156],
        [-0.1579, -0.1254, -0.1588, -0.1136],
        [-0.0484, -0.0718, -0.1674, -0.1839]])
```

### the current code
```
grads: (tensor([[-0.0576, -0.0557, -0.0831, -0.0729],
        [-0.0867, -0.0694, -0.0778, -0.0508],
        [-0.0852, -0.0687, -0.0781, -0.0520],
        [ 0.0289, -0.0130, -0.0879, -0.1265]]), tensor([[0.0354, 0.0653, 0.0552, 0.0957],
        [0.0354, 0.0651, 0.0548, 0.0950],
        [0.0353, 0.0649, 0.0546, 0.0946],
        [0.0352, 0.0647, 0.0542, 0.0939]]))
grads: (tensor([[0.0454, 0.0753, 0.0521, 0.0859],
        [0.0457, 0.0759, 0.0524, 0.0865],
        [0.0457, 0.0759, 0.0524, 0.0865],
        [0.0393, 0.0649, 0.0451, 0.0740]]), tensor([[-0.0629, -0.0525, -0.0833, -0.0704],
        [-0.0692, -0.0552, -0.0817, -0.0648],
        [-0.0726, -0.0567, -0.0807, -0.0617],
        [-0.0773, -0.0587, -0.0794, -0.0574]]))
all_gradients after reduce: tensor([[[-0.0121,  0.0196, -0.0311,  0.0130],
         [-0.0410,  0.0065, -0.0254,  0.0357],
         [-0.0395,  0.0072, -0.0257,  0.0346],
         [ 0.0682,  0.0518, -0.0428, -0.0525]],

        [[-0.0275,  0.0129, -0.0281,  0.0253],
         [-0.0339,  0.0099, -0.0268,  0.0303],
         [-0.0373,  0.0083, -0.0261,  0.0329],
         [-0.0421,  0.0059, -0.0252,  0.0365]]])
all_gradients after reduce: tensor([[[-0.0121,  0.0196, -0.0311,  0.0130],
         [-0.0410,  0.0065, -0.0254,  0.0357],
         [-0.0395,  0.0072, -0.0257,  0.0346],
         [ 0.0682,  0.0518, -0.0428, -0.0525]],

        [[-0.0275,  0.0129, -0.0281,  0.0253],
         [-0.0339,  0.0099, -0.0268,  0.0303],
         [-0.0373,  0.0083, -0.0261,  0.0329],
         [-0.0421,  0.0059, -0.0252,  0.0365]]])
grad_out: tensor([[-0.0121,  0.0196, -0.0311,  0.0130],
        [-0.0410,  0.0065, -0.0254,  0.0357],
        [-0.0395,  0.0072, -0.0257,  0.0346],
        [ 0.0682,  0.0518, -0.0428, -0.0525]])
grad_out: tensor([[-0.0275,  0.0129, -0.0281,  0.0253],
        [-0.0339,  0.0099, -0.0268,  0.0303],
        [-0.0373,  0.0083, -0.0261,  0.0329],
        [-0.0421,  0.0059, -0.0252,  0.0365]])
```

Comment (ordabayevy) [reply lightly/utils/dist.py:23]: You are right, thanks! Each gradient in grads needs to be reduced across all ranks.

I initially had the code below that was working but then switched to the one above without properly testing it with the model.

```py
    @staticmethod
    def backward(ctx, *grads) -> torch.Tensor:
        new_grads = []
        for grad in grads:
            grad = grad.contiguous()
            dist.all_reduce(grad, op=dist.ReduceOp.SUM)
            new_grads.append(grad)
        grad_out = new_grads[dist.get_rank()]
        return grad_out
```

Comment (MalteEbner) [reply lightly/utils/dist.py:23]: Both of the methods you proposed @ordabayevy give the same results:
```
    def backward(ctx, *grads) -> torch.Tensor: 
        new_grads = []
        for grad in grads:
            grad = grad.contiguous()
            dist.all_reduce(grad, op=dist.ReduceOp.SUM)
            new_grads.append(grad)
        print(f"rank: {dist.get_rank()}, new_grads: {new_grads}")
        grad_out = new_grads[dist.get_rank()]
        print(f"rank: {dist.get_rank()}, grad_out: {grad_out}")
        return grad_out
```
```
rank: 0, new_grads: [tensor([[-0.0121,  0.0196, -0.0311,  0.0130],
        [-0.0410,  0.0065, -0.0254,  0.0357],
        [-0.0395,  0.0072, -0.0257,  0.0346],
        [ 0.0682,  0.0518, -0.0428, -0.0525]]), tensor([[-0.0275,  0.0129, -0.0281,  0.0253],
        [-0.0339,  0.0099, -0.0268,  0.0303],
        [-0.0373,  0.0083, -0.0261,  0.0329],
        [-0.0421,  0.0059, -0.0252,  0.0365]])]rank: 1, new_grads: [tensor([[-0.0121,  0.0196, -0.0311,  0.0130],
        [-0.0410,  0.0065, -0.0254,  0.0357],
        [-0.0395,  0.0072, -0.0257,  0.0346],
        [ 0.0682,  0.0518, -0.0428, -0.0525]]), tensor([[-0.0275,  0.0129, -0.0281,  0.0253],
        [-0.0339,  0.0099, -0.0268,  0.0303],
        [-0.0373,  0.0083, -0.0261,  0.0329],
        [-0.0421,  0.0059, -0.0252,  0.0365]])]

rank: 1, grad_out: tensor([[-0.0275,  0.0129, -0.0281,  0.0253],
        [-0.0339,  0.0099, -0.0268,  0.0303],
        [-0.0373,  0.0083, -0.0261,  0.0329],
        [-0.0421,  0.0059, -0.0252,  0.0365]])
rank: 0, grad_out: tensor([[-0.0121,  0.0196, -0.0311,  0.0130],
        [-0.0410,  0.0065, -0.0254,  0.0357],
        [-0.0395,  0.0072, -0.0257,  0.0346],
        [ 0.0682,  0.0518, -0.0428, -0.0525]])
```

Comment (MalteEbner) [reply lightly/utils/dist.py:24]: Resolving this to allow merging
